mesos-client/src/main/scala/mesosphere/mesos/client/MesosClient.scala

```diff
@@ -0,0 +1,514 @@
+
+    val recordIoScanner: Flow[ByteString, ByteString, NotUsed] = RecordIOFraming.scanner()
+
+    val connectionHandler: Flow[HttpResponse, HttpResponse, NotUsed] = Flow[HttpResponse].map { response =>
+      response.status match {
+        case StatusCodes.OK =>
+          logger.info(s"Connected successfully to ${context.get().url}");
+          val streamId = response.headers
+            .find(h => h.is(MesosStreamIdHeaderName.toLowerCase))
+            .getOrElse(throw new IllegalStateException(s"Missing MesosStreamId header in ${response.headers}"))
+
+          // At this point we successfully connected to mesos leader so context should have correct leader's host
+          // and port either from the config or set on the redirect.
+          context.updateAndGet(c => c.copy(mesosStreamId = Some(streamId.value())))
+          response
+        case StatusCodes.TemporaryRedirect =>
+          val leader = new URI(response.header[headers.Location].get.value())
+          logger.warn(s"New mesos leader available at $leader")
+          // Update the context with the new leader's host and port and throw an exception that is handled in the
+          // next `recoverWith` stage.
+          context.updateAndGet(c => c.copy(host = leader.getHost, port = leader.getPort))
+          throw new MesosRedicrectException(leader)
+        case _ =>
+          throw new IllegalArgumentException(s"Mesos server error: $response")
+      }
+    }
+
+    val dataBytesExtractor: Flow[HttpResponse, ByteString, NotUsed] = Flow[HttpResponse].flatMapConcat(resp => resp.entity.dataBytes)
+
+    val eventDeserializer: Flow[ByteString, Event, NotUsed] = Flow[ByteString].map(bytes => Event.parseFrom(bytes.toArray))
+
+    val subscribedHandler: Flow[Event, Event, NotUsed] = Flow[Event].map { event =>
+      if (event.subscribed.isDefined) {
+        val ctx = context.updateAndGet(c => c.copy(frameworkId = Some(event.subscribed.get.frameworkId)))
+        // We save `frameworkId` in the context and successfully complete the promise, meaning the calls in sink
+        contextPromise.success(ctx)
+      }
+      event
+    }
+
+    def connectionSource(host: String, port: Int) =
+      Source.single(request)
+      .via(log(s"Connecting to the new leader: $host:$port"))
+      .via(httpConnection(host, port))
+      .via(log("HttpResponse: "))
+      .via(connectionHandler)
+
+    connectionSource(context.get().host, context.get().port)
+      .recoverWithRetries(conf.redirectRetires(), {
+        case MesosRedicrectException(_) => connectionSource(context.get().host, context.get().port)
+      })
+      .idleTimeout(conf.idleTimeout().seconds)
+      .buffer(conf.sourceBufferSize(), overflowStrategy)
+      .via(dataBytesExtractor)
+      .via(recordIoScanner)
+      .via(eventDeserializer)
+      .via(subscribedHandler)
+      .via(log("Received mesos Event: "))
+  }
+
+  override lazy val (killSwitch: UniqueKillSwitch, mesosSource: Source[Event, NotUsed]) = source
```

Comment:
as seen above ^ I'm not sure that using Broadcast hub by default is a good idea.

Perhaps we can just expose the `Source`, and then leave it up to the user to use BroadcastHub or not? And, perhaps we should just connect to Mesos when the user materializes said Source?

Reply:
+1

Reply:
-1 on that. Main reason why I inserted broadcast huyb was that otherwise a new mesos connection (SUBSCRIBE) would be established on every `mesosSource` materialization. That is not what we want. We want 1 connection in (and multiple requests out). But may be there is a better way to achieve this (without using a `BroadcastHub`)?

Reply:
See here: https://doc.akka.io/docs/akka-http/current/client-side/connection-level.html#opening-http-connections

Reply:
Can we not just make that clear in the documentation? "A new Mesos connection will be established for every source materialization. If you require multiple consumers, please consider using BroadcastHub"

There are multiple other ways to avoid using BroadcastHub. You can use the Akka stream graph DSL, you can use the alsoTo DSL operator on flows/sources, or you can use some other broadcast mechanism.

I think the ideal is that the user of this library makes the decision that makes the most sense. Personally, I would not choose to use BroadcastHub because I lose the guarantee that might consumers will receive the first messages.